internal/handler/control.go

```diff
@@ -1,22 +1,142 @@
 // -*- Mode: Go; indent-tabs-mode: t -*-
 //
 // Copyright (C) 2017-2018 Canonical Ltd
+// Copyright (C) 2020 IOTech Ltd
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package handler
 
 import (
+	"context"
 	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"sync/atomic"
+	"time"
 
+	"github.com/edgexfoundry/device-sdk-go/internal/cache"
 	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/pkg/models"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+	"github.com/google/uuid"
 )
 
-func DiscoveryHandler(requestMap map[string]string) {
-	common.LoggingClient.Info(fmt.Sprintf("service: discovery request"))
-}
+var (
+	id     string
+	locker uint32
+)
 
 func TransformHandler(requestMap map[string]string) (map[string]string, common.AppError) {
 	common.LoggingClient.Info(fmt.Sprintf("service: transform request: transformData: %s", requestMap["transformData"]))
 	return requestMap, nil
 }
+
+func DiscoveryHandler(w http.ResponseWriter) {
+	if id == "" {
+		id = uuid.New().String()
+	}
+
+	if w != nil {
+		msg := fmt.Sprintf("Discovery triggered or already running, id = %s", id)
+		w.WriteHeader(http.StatusAccepted) //status=202
+		_, _ = io.WriteString(w, msg)
+	}
+
+	// locker is used to prevent duplicate call of protocol specific device discovery
+	// locker would be released once the DS main thread received the candidate devices
+	// found by discovery
+	if !atomic.CompareAndSwapUint32(&locker, 0, 1) {
+		common.LoggingClient.Info(fmt.Sprintf("Discovery request returned. discovery process is running"))
+		return
+	}
+	common.LoggingClient.Info(fmt.Sprintf("service %s discovery triggered", common.ServiceName))
+
+	ctx := context.WithValue(context.Background(), common.CorrelationHeader, id)
+	deviceCh := make(chan []models.DiscoveredDevice)
+	go common.Discovery.Discover(deviceCh)
+	go filterAndAddition(ctx, deviceCh)
+}
+
+func filterAndAddition(ctx context.Context, deviceCh <-chan []models.DiscoveredDevice) {
+	pws := cache.ProvisionWatchers().All()
+	devices := <-deviceCh
+	// Notify the DS main thread to accept discovery request
+	atomic.StoreUint32(&locker, 0)
+	id = ""
+
+	for _, d := range devices {
+		for _, pw := range pws {
+			if !whitelistPass(d, pw) {
+				break
+			}
+			if !blacklistPass(d, pw) {
+				break
+			}
+
+			if _, ok := cache.Devices().ForName(d.Name); ok {
+				common.LoggingClient.Info(fmt.Sprintf("Candidate discovered device %s already existed", d.Name))
```

Comment:
this should be a Debug log statement, otherwise the more devices that get added via discovery cause more and more log spam